### PR TITLE
Angular. Components. Directives & Pipes

### PR DESCRIPTION
**Pull request by aleksander1802**
1. Task: [link](https://github.com/rolling-scopes-school/tasks/blob/master/tasks/angular/components-directives-pipes.md)
2. Screenshot: 
localhost_4200_ (1)
3. Deploy: [link](https://aleksander1802.github.io/angular2023Q4/)
4. Done: 28.10.2023 / Deadline: 31.10.2023
5. Score: 100 / 100

## Scoring: Angular. Components. Directives & Pipes

- [x] Main layout is implemented (**+10**)
- [x] Fake search functionality is implemented: by submitting the form in the **Header** component, search results are shown (**+20**)
- [x] The **Search result item** component contains all the necessary data (**+10**)
- [x] The **Filtering criteria block** toggle functionality is implemented (**+5**)
- [x] The colored border under the **Search result item** is implemented using a `Directive` (**+15**)
- [x] Search result sorting is implemented (both ascending and descending directions) (**+15**)
- [x] Search result filtering by key words is implemented using a `Pipe` (**+15**)
- [x] Custom Button component is implemented and used across the application. The component utilizes [Content projection](https://angular.io/guide/content-projection) to display the button text (**+5**)
- [x] One of the components is implemented as [standalone](https://angular.io/guide/standalone-components). Strong recommendation to choose a simple presentational component and not a "Smart" container component for this task (**+5**)

### CODE REVIEW

File: src/app/shared/components/custom-button/custom-button.component.html [hunk 1]
```diff
@@ -0,0 +1,3 @@
+<button [ngClass]="defaultClasses + ' ' + additionalClasses">
```
Comment (ShkrobNastya) [inline src/app/shared/components/custom-button/custom-button.component.html]: if you sure that custom-button class will be always here you can left it in html and remove variable in ts file

File: src/app/shared/components/logo/logo.component.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<div class="logo" routerLink="/">
+    <img src="assets/images/youtube.svg" alt="youtube logo" />
+</div>
```
Comment (ShkrobNastya) [inline src/app/shared/components/logo/logo.component.html:3]: **you don't need to fix it**

not sure that in the scope of this project we need to create a separate component for a logo. But if it is a big project with different configurations (for ex. add title for some cases) it makes sense

File: src/app/services/results/results.service.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+export class ResultsService {
+    isResultsVisible = false;
+    private mockResponse: SearchResultResponse;
+    private mockItems: SearchItem[];
+    private searchResultsSubject = new BehaviorSubject<SearchItem[]>([]);
+    public searchResults = this.searchResultsSubject.asObservable();
+
+    constructor() {
+        this.mockResponse = mockResponse;
+        this.mockItems = mockResponse.items;
```
Comment (ShkrobNastya) [inline src/app/services/results/results.service.ts]: please, don't use constructor for such cases) 
in angular constructor mostly used for declaring services and call super method like this 
 constructor(private resultsService: ResultsService) {}

for such purposes we use ngOnInit hook method

File: src/app/components/filtering-criteria-block/filtering-criteria-block.component.html [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+                {{
+                    sortService.sortOrder === "asc"
+                        ? "arrow_upward"
+                        : "arrow_downward"
+                }}
+            </mat-icon>
+        </a>
+        <a
+            href="#"
+            (click)="sortService.sortByViews($event); setActiveSorting('views')"
```
Comment (ShkrobNastya) [inline src/app/components/filtering-criteria-block/filtering-criteria-block.component.html]: it is better to create a separate method in component fot click event

File: src/app/components/filtering-criteria-block/filtering-criteria-block.component.html [hunk 5]
```diff
@@ -0,0 +1,42 @@
+<div class="header__container-sort">
+    <span>Sorting by:</span>
+    <div class="sort">
+        <a
+            href="#"
+            (click)="sortService.sortByDate($event); setActiveSorting('date')"
```
Comment (ShkrobNastya) [inline src/app/components/filtering-criteria-block/filtering-criteria-block.component.html]: it is better to create a separate method in component fot click event

**Review by ShkrobNastya — CHANGES_REQUESTED** [review]

File: src/app/shared/components/custom-button/custom-button.component.html (hunk 1 above)
Comment (aleksander1802) [reply src/app/shared/components/custom-button/custom-button.component.html]:
```
<button class="custom-button" [ngClass]="additionalClasses">
    <ng-content></ng-content>
</button>
```
Fixed!

File: src/app/services/results/results.service.ts (hunk 3 above)
Comment (aleksander1802) [reply src/app/services/results/results.service.ts]:
```
    isResultsVisible = false;
    private mockResponse: SearchResultResponse = mockResponse;
    private mockItems: SearchItem[] = this.mockResponse.items;
    private searchResultsSubject = new BehaviorSubject<SearchItem[]>([]);
    public searchResults = this.searchResultsSubject.asObservable();
```
Fixed!

File: src/app/components/filtering-criteria-block/filtering-criteria-block.component.html (hunk 4 above)
Comment (aleksander1802) [reply src/app/components/filtering-criteria-block/filtering-criteria-block.component.html]:
```
<a
            href="#"
            (click)="handleSortByViews($event)"
            [ngClass]="getLinkClass('views')"
        >
            count of views
            <mat-icon *ngIf="sortService.sortBy === 'views'">
                {{
                    sortService.sortOrder === "asc"
                        ? "arrow_upward"
                        : "arrow_downward"
                }}
            </mat-icon>
        </a>
```

```
handleSortByViews(event: Event) {
        this.sortService.sortByViews(event);
        this.setActiveSorting('views');
    }
```

Fixed!

File: src/app/components/filtering-criteria-block/filtering-criteria-block.component.html (hunk 5 above)
Comment (aleksander1802) [reply src/app/components/filtering-criteria-block/filtering-criteria-block.component.html]:
```
<a
            href="#"
            (click)="handleSortByDate($event)"
            [ngClass]="getLinkClass('date')"
        >
            date
            <mat-icon *ngIf="sortService.sortBy === 'date'">
                {{
                    sortService.sortOrder === "asc"
                        ? "arrow_upward"
                        : "arrow_downward"
                }}
            </mat-icon>
        </a>
```
```
handleSortByDate(event: Event) {
        this.sortService.sortByDate(event);
        this.setActiveSorting('date');
    }
```

Fixed!

**Review by ShkrobNastya — APPROVED** [review]

**Pull request merged**